### PR TITLE
Build eglib using the libil2cpp OS API

### PR DESCRIPTION
**Pull request by joshpeterson**
This pull request adds a new static library named eglib-unity. This static library is the same as eglib, but with parts of the OS API replaced by the Unity equivalent from libil2cpp.

* This change includes only `g_get_tmp_dir` from the Unity OS API, as a proof of concept.
* The new eglib-unity will be used only for IL2CPP on Mono at this time.
* The builds for libmonoruntime-il2cpp and libmonoutils-il2cpp will now be done with a different target in the Makefile.
* Changes to the monobuildtools repo to go with this change are on the [unity-mono-4.8.0-branch-os-layer](https://github.com/Unity-Technologies/monobuildtools/tree/unity-mono-4.8.0-branch-os-layer) branch. I'll merge those changes first, then update the submodule in this branch before this PR is merged.
* The name of the eglib static library will change now, so I'll need to land changes in the IL2CPP repo to account for that when the new Mono binaries land in il2cpp-dependencies.

A few things to consider:

* Do we want to land this on the 4.8 branch of Mono (as this PR does)? Should we wait until we have a Mono 5.0 branch?
* This makes the builds for OSX more invasive, as the Makefiles are now changed. I think that the builds are cleaner overall though. Is this worth the trade-off?

### CODE REVIEW

File: mono/metadata/Makefile.am [hunk 1]
```diff
@@ -70,10 +70,10 @@ endif
 
 if SUPPORT_BDWGC
 if DISABLE_EXECUTABLES
-shared_bdwgc_libraries = libmonoruntimebdwgc.la
+shared_bdwgc_libraries = libmonoruntimebdwgc.la libmonoruntime-il2cpp-bdwgc.la
```
Comment (joncham) [inline mono/metadata/Makefile.am]: Are these changes needed for this PR? Or for future work?

File: eglib/src/gmisc-unity.c [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+#include "Path-c-api.h"
+
+static const char *tmp_dir;
+
+const gchar *
+g_get_tmp_dir(void)
+{
+    if (tmp_dir == NULL)
+        tmp_dir = UnityPalGetTempPath();
```
Comment (mderoy) [inline eglib/src/gmisc-unity.c]: I assume our UnityPalGetTempPath allocates memory...wondering if this should be
char* path = UnityPalGetTempPath();
tmp_dir = strdup(path);
delete path

I'm worried about mono calling g_free on strings that we hand it from il2cpp's allocator (though maybe those worries are unfounded?)

Comment (joshpeterson) [inline eglib/src/gmisc-unity.c]: We're actually giving the libil2cpp OS layer Mono's allocator, so the the memory is allocated in the same way the Mono code we are replacing allocates it.

File: mono/metadata/Makefile.am (hunk 1 above)
Comment (mderoy) [inline mono/metadata/Makefile.am]: I think Josh made libmonoruntime-il2cpp-bdwgc a separate built target so we don't need to call make an additional time

Comment (joshpeterson) [inline mono/metadata/Makefile.am]: Yes, that is correct!

File: msvc/unity.props [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <PLATFORM_UNITY_DEFINES>PLATFORM_UNITY</PLATFORM_UNITY_DEFINES>
+    <PLATFORM_UNITY_INCLUDE_DIR>../../../il2cpp/build/libil2cpp/os/c-api</PLATFORM_UNITY_INCLUDE_DIR>
```
Comment (mderoy) [inline msvc/unity.props]: I think we need to figure out a solution to this circular dependency..maybe move the il2cpp platform abstraction layer into github?

File: mono/metadata/Makefile.am (hunk 1 above)
Comment (mderoy) [inline mono/metadata/Makefile.am]: Wheres the buildscripts PR ;)

**Comment by mderoy** [comment]
@joshpeterson looks like you're updating external buildscripts in here, but I think you need to push those changes first

File: msvc/unity.props (hunk 3 above)
Comment (joshpeterson) [inline msvc/unity.props]: I agree. It is not strictly a circular dependency, as Mono only depends on the OS layer in libil2cpp, and nothing in that code depends on Mono. For now, I think this is the best option, but in the future we'll pull the OS code out of libil2cpp into a separate library or repo and reference that.

**Comment by joshpeterson** [comment]
@mderoy Yes, I am changing the monobuildtools repo as we well. Changes to the monobuildtools repo to go with this change are on the unity-mono-4.8.0-branch-os-layer branch. I'll merge those changes first, then update the submodule in this branch before this PR is merged. I was not planning a separate PR in the monobuildtools repo, as you can click through the links in this PR to see the diff of the changes in that repo. But I can do a separate PR there if you would like. What is your opinion?

File: mono/metadata/Makefile.am (hunk 1 above)
Comment (joshpeterson) [inline mono/metadata/Makefile.am]: See my comment below.

**Comment by mderoy** [comment]
@joshpeterson I'll take a look, but I didn't notice that in my initial review, maybe just a note in the comment in the future so we know to look for the submodule changes

**Comment by joshpeterson** [comment]
Sounds good. I'll try to make things more clear about the submodule changes in the future.

**Pull request merged**